File: 011_Container With Most Water/Container With Most Water.py
"""
LeetCode 11
给你 n 个非负整数 a1，a2，...，an，每个数代表坐标中的一个点 (i, ai)。
在坐标内画 n 条垂直线，垂直线 i 的两个端点分别为 (i, ai) 和 (i, 0)。
找出其中的两条线，使得它们与 x 轴共同构成的容器可以容纳最多的水。
说明：你不能倾斜容器，且 n 的值至少为 2。
示例：
    输入：[1,8,6,2,5,4,8,3,7]
    输出：49
"""


class Solution:
    """ 2020/4/5 双指针 68ms 15.2MB """

    @staticmethod
    def max_area(height: list) -> int:
        if len(height) == 0:
            return 0
        ptr_l = 0
        ptr_r = len(height) - 1
        max_area = min(height[ptr_l], height[ptr_r]) * (ptr_r - ptr_l)
        while ptr_l < ptr_r:
            if height[ptr_l] < height[ptr_r]:
                ptr_l = ptr_l + 1
            else:
                ptr_r = ptr_r - 1
            if max_area < min(height[ptr_l], height[ptr_r]) * (ptr_r - ptr_l):
                max_area = min(height[ptr_l], height[ptr_r]) * (ptr_r - ptr_l)
        return max_area


if __name__ == '__main__':
    height_test = [1, 8, 6, 2, 5, 4, 8, 3, 7]
    print(Solution.max_area(height_test))

File: 0011_container_with_most_water/Container With Most Water.c
//
// Created by david on 2020/4/5.
//

#include <stdio.h>

#define SOLUTION 2

#if SOLUTION == 1

/* 20020/4/5 双指针 8ms 6.4MB */
int maxArea(const int *height, int heightSize) {
    if (heightSize < 2)
        return 0;

    int *ptr_l = (int *) height;
    int *ptr_r = (int *) height + heightSize - 1;
    int maxValue = (int) (ptr_r - ptr_l) * ((*ptr_l < *ptr_r) ? *ptr_l : *ptr_r);

    while (ptr_r != ptr_l) {
        if (*ptr_l <= *ptr_r) ptr_l++;
        else ptr_r--;
        maxValue = (maxValue > (int) (ptr_r - ptr_l) * ((*ptr_l < *ptr_r) ? *ptr_l : *ptr_r)) ?
                   maxValue : (int) (ptr_r - ptr_l) * ((*ptr_l < *ptr_r) ? *ptr_l : *ptr_r);
    }
    return maxValue;
}

#elif SOLUTION == 2

/* 20020/4/5 双指针 */
int maxArea(const int *height, int heightSize) {
    int temp_area;
    int maxValue = 0;
    int ptr_l = 0;
    int ptr_r = heightSize - 1;

    while (ptr_l < ptr_r) {
        if (height[ptr_l] < height[ptr_r]) {
            temp_area = height[ptr_l] * (ptr_r - ptr_l);
            ptr_l++;
        } else {
            temp_area = height[ptr_r] * (ptr_r - ptr_l);
            ptr_r--;
        }

        maxValue = maxValue < temp_area ? temp_area : maxValue;
    }
    return maxValue;
}

#endif

void maxAreaTest(void) {
    int height[] = {1, 1};
    printf("%d\n", maxArea(height, sizeof(height) / sizeof(height[0])));
}

int main() {
    maxAreaTest();
    return 0;
}